File: MainApp/src/main/java/com/sleticalboy/util/ThreadHelper.java
package com.sleticalboy.util;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;

/**
 * Created on 20-4-2.
 *
 * @author binlee sleticalboy@gmail.com
 */
public final class ThreadHelper {

    private static Handler sMain;
    private static Handler sWorker;

    private ThreadHelper() {
        //no instance
    }

    private static void ensureMain() {
        if (sMain != null) {
            return;
        }
        sMain = new Handler(Looper.getMainLooper());
    }

    private static void ensureWorker() {
        if (sWorker != null) {
            return;
        }
        final HandlerThread thread = new HandlerThread("ThreadHelper");
        thread.start();
        sWorker = new Handler(thread.getLooper());
    }

    private static void exec(boolean main, Runnable task, long delay) {
        if (task == null) {
            return;
        }
        if (main) {
            if (Looper.getMainLooper() == Looper.myLooper()) {
                task.run();
                return;
            }
            ensureMain();
        } else {
            ensureWorker();
        }
        (main ? sMain : sWorker).postDelayed(task, delay);
    }

    public static void runOnMain(Runnable task) {
        exec(true, task, -1);
    }

    public static void runOnMain(Runnable task, long delay) {
        exec(true, task, delay);
    }

    public static void runOnWorker(Runnable task) {
        exec(false, task, -1);
    }

    public static void runOnWorker(Runnable task, long delay) {
        exec(false, task, delay);
    }

}

File: MainApp/src/main/java/com/sleticalboy/util/ThreadHelper.kt
package com.binlee.util

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper

/**
 * Created on 20-4-2.
 *
 * @author binlee sleticalboy@gmail.com
 */
object ThreadHelper {

  private var sMain: Handler? = null
  private var sWorker: Handler? = null

  private fun ensureMain() {
    if (sMain != null) return
    sMain = Handler(Looper.getMainLooper())
  }

  private fun ensureWorker() {
    if (sWorker != null) return
    val thread = HandlerThread("ThreadHelper")
    thread.start()
    sWorker = Handler(thread.looper)
  }

  private fun exec(main: Boolean, task: Runnable?, delay: Long) {
    if (task == null) return
    if (main) {
      if (Looper.getMainLooper() == Looper.myLooper()) {
        task.run()
        return
      }
      ensureMain()
    } else {
      ensureWorker()
    }
    (if (main) sMain else sWorker)!!.postDelayed(task, delay)
  }

  fun runOnMain(task: Runnable?) {
    exec(true, task, -1)
  }

  fun runOnMain(task: Runnable?, delay: Long) {
    exec(true, task, delay)
  }

  fun runOnWorker(task: Runnable?) {
    exec(false, task, -1)
  }

  fun runOnWorker(task: Runnable?, delay: Long) {
    exec(false, task, delay)
  }
}